subsys/suit/stream/stream_sinks/src/suit_sdfw_recovery_sink.c

```diff
@@ -0,0 +1,366 @@
+	};
+
+	LOG_DBG("update_candidate: 0x%08lX", (uintptr_t)buf);
+	LOG_DBG("signed_manifest: 0x%08lX", update_blob.manifest_addr);
+	LOG_DBG("public_key: 0x%08lX", update_blob.pubkey_addr);
+	LOG_DBG("signature: 0x%08lX", update_blob.signature_addr);
+	LOG_DBG("firmware: 0x%08lX", update_blob.firmware_addr);
+	LOG_DBG("max update size: 0x%08X", update_blob.max_size);
+	LOG_DBG("firmware size: 0x%08X (%d)", size, size);
+
+	/* NOTE: SecROM does not use the actual size of new SDFW during update process.
+	 *       However, if SDFW is flashed for the first time, the value is used to limit the
+	 * future maximum SDFW size. Hence, use the maximum allowed size for safety.
+	 */
+	err = sdfw_recovery_update(&update_blob);
+
+	if (err) {
+		LOG_ERR("Failed to schedule SDFW Recovery update: %d", err);
+		return SUIT_PLAT_ERR_CRASH;
+	}
+
+	LOG_INF("SDFW Recovery update scheduled");
+
+	return SUIT_PLAT_SUCCESS;
+}
+
+static sdf_sink_err_t check_update_candidate(const uint8_t *buf, size_t size)
+{
+	uint8_t *candidate_binary_start =
+		(uint8_t *)(buf + CONFIG_SUIT_SDFW_RECOVERY_UPDATE_FIRMWARE_OFFSET);
+	uint8_t *candidate_digest_in_manifest =
+		(uint8_t *)(buf + CONFIG_SUIT_SDFW_RECOVERY_UPDATE_DIGEST_OFFSET);
+	uint8_t *current_sdfw_recovery_digest =
+		(uint8_t *)(NRF_SICR->UROT.RECOVERY.SM.TBS.FW.DIGEST);
+
+	/* First check if calculated digest of candidate matches the digest from Signed Manifest */
+	digest_sink_err_t err =
+		verify_digest(candidate_binary_start,
+			      size - (size_t)CONFIG_SUIT_SDFW_RECOVERY_UPDATE_FIRMWARE_OFFSET,
+			      PSA_ALG_SHA_512, candidate_digest_in_manifest);
+
+	if (err != SUIT_PLAT_SUCCESS) {
+		if (err == DIGEST_SINK_ERR_DIGEST_MISMATCH) {
+			LOG_ERR("Candidate inconsistent");
+		} else {
+			LOG_ERR("Failed to calculate digest: %d", err);
+		}
+
+		return SUIT_PLAT_ERR_CRASH;
+	}
+
+	LOG_DBG("Candidate consistent");
+
+	/* Then compare candidate's digest with current SDFW Recovery digest */
+	err = verify_digest(candidate_binary_start,
+			    size - (size_t)CONFIG_SUIT_SDFW_RECOVERY_UPDATE_FIRMWARE_OFFSET,
+			    PSA_ALG_SHA_512, current_sdfw_recovery_digest);
+	if (err == SUIT_PLAT_SUCCESS) {
+		LOG_INF("Same candidate - skip update");
+		return SUIT_PLAT_SUCCESS;
+	} else if (err == DIGEST_SINK_ERR_DIGEST_MISMATCH) {
+		LOG_INF("Different candidate");
+		err = schedule_sdfw_recovery_update(buf, size);
+		if (err == SUIT_PLAT_SUCCESS) {
+			LOG_DBG("Update scheduled");
+			err = SDFW_RECOVERY_SINK_ERR_AGAIN;
+		}
+		return err;
+	}
+
+	LOG_ERR("Failed to calculate digest: %d", err);
+	return SUIT_PLAT_ERR_CRASH;
+}
+
+static void reboot_to_continue(void)
+{
+	if (IS_ENABLED(CONFIG_SUIT_UPDATE_REBOOT_ENABLED)) {
+		LOG_INF("Reboot the system to continue SDFW Recovery update");
+
+		LOG_PANIC();
+
+		sys_reboot(SYS_REBOOT_COLD);
+	} else {
+		LOG_DBG("Reboot disabled - perform manually");
+	}
+}
+
+static suit_plat_err_t check_urot_none(const uint8_t *buf, size_t size)
```

Comment:
Is `urot` the correct name in case of recovery?

Reply:
Yes, this is a name of the register, which is used for both sdfw and sdfw recovery updates.